rfcs/001-managed-ptr-type.md

```diff
@@ -0,0 +1,94 @@
+- Topic Name: managed_ptr_type
+- Start Date: 2022-03-29
+- RFC PR: [vlang/rfcs#23](https://github.com/vlang/rfcs/pull/23)
+
+# Summary
+
+This RFC is about annotating pointer types as appearing in function parameters and return values as well as struct members with a hint as to wether they are pointing to a heap object, or a stack object or member of a heap object. This is important since ideally we want to manage the lifetime of heap objects automatically through a combination of auto-free and reference counting. And while the compiler can infer all information it needs inside a function it cannot do so across functions or structs, hence the need arises for such an annotation.
+The goal here is to strike a unique balance between poles like C/C++, Lobster, Rust, and others that fot the unique needs of of being easy to use, safe, but also allows some of the flexibility pointers give in C/C++ and other languages.
+
+# Motivation
+
+Implementing this proposal will enable us to automatically manage heap objects leaving and entering functions and living inside of structs, and thus prevent memory and resource leaks and free the programmer from manual resource/memory management.
+
+# Guide-level explanation
+
+This proposal introduces an alternative pointer syntax for function return values:
+
+```
+fn foo() ^Struct
+{
+    return &Struct{}
+}
+
+```
+
+struct members:
+
+```
+struct Struct
+{
+    other ^Struct
+};
+```
+
+ and function arguments:
+
+ ```
+ fn (s &Struct)(other ^Struct)
+ {
+    s.other = other
+ }
+ ```
+
+or even
+
+ ```
+ fn (s ^Struct)(other &Struct)
+ {
+    other.other = other
+ }
+ ```
+
+ where replacing `&` with `^` in struct pointers used as function arguments, returns or struct members signify that these are managed heap objects and the thusly annotated objects transport (possibly shared) ownership of these objects.
```

Comment:
Why the `^` and maybe not the `'`?

The operator `'` is common already in a language like Rust (with different meanings), but maybe avoid confusion between devs?

Reply:
I am not married to `^`. I think I have seen it used elsewhere for pointer like types. Also I think `'`is a little hard to see? But as I said, I don't care too much. The principal question is do we want a distinct managed pointer type (and I think we do because the alternatives look grim...)

Reply:
As someone mentioned on discord, `^` is used in Odin.

Personally, I would still prefer something be done so no extra annotation was required.  The computer has a lot more cycles to figure things out than I want to spend worrying about such details.

If the only thing missing is knowing whether something is on the heap or not, perhaps an extra boolean in object structures... `heap: true`?

Reply:
> If the only thing missing is knowing whether something is on the heap or not, perhaps an extra boolean in object structures... `heap: true`?

what do you mean? in the struct definition? internally the compiler would do just that. but it has to know what to put into that boolean

Reply:
Yes, in the struct definition.  And it is known at the time the object is created, otherwise it couldn't be created at all.

Reply:
Ah, now I get what you mean. You mean effectively move the annotation to run time. Hmm, let me think about this...

Reply:
**Any** tagging shall be unified with sum types optimizations (especially the most common `sometype | none` sum type). Otherwise it'll get unwieldy and thus error-prone.

Still, I'd also prefer compile-time solution due to all the guarantees.

Reply:
Btw, in case we opt for runtime tagging we can always use static analysis to find the dangling references, so it would not be out of the picture

Reply:
That's the question of choice here. I feel it makes sense to express such things explicitly, but i can also accept that this is not the choice for this language. How would such matters be resolved? Is there a kind of voting process or should we leave it up to alex? I am ok either way to be honest.

Reply:
Alex makes the final decision (since it's his language), but we are free to give whatever input we can.  The better the input (well-reasoned arguments, all the way up to PRs with benchmarks), the more chance he'll vote your way.  :-)